File: AC/ACL/acl001c.go
package main

import "fmt"

type unionFind struct {
	count   int
	parents []int
}

func NewUnionFind(n int) *unionFind {
	p := make([]int, n)
	for i := 0; i < n; i++ {
		p[i] = i
	}
	return &unionFind{
		count:   n,
		parents: p,
	}
}

func (u *unionFind) Root(x int) int {
	if u.parents[x] == x {
		return x
	}
	u.parents[x] = u.Root(u.parents[x])
	return u.parents[x]
}

func (u *unionFind) Unite(x, y int) {
	rx, ry := u.Root(x), u.Root(y)
	switch {
	case rx == ry:
		return
	case rx < ry:
		u.parents[rx] = ry
	case rx > ry:
		u.parents[ry] = rx
	}
}

func (u *unionFind) Same(x, y int) bool {
	return u.Root(x) == u.Root(y)
}

func main() {
	var n, m, a, b int
	fmt.Scan(&n, &m)
	var cnt = n
	u := NewUnionFind(n)
	for i := 0; i < m; i++ {
		fmt.Scan(&a, &b)
		if u.Same(a-1, b-1) {
			continue
		} else {
			u.Unite(a-1, b-1)
			cnt--
		}
	}
	if cnt > 0 {
		fmt.Println(cnt - 1)
	} else {
		fmt.Println(cnt)
	}
}

File: AC/ACL/acl001c.py
class UnionFind():
    def __init__(self, n):
        self.n = n
        self.parents = [-1]*n
        for i in range(n):
            self.parents[i] = i

    # 要素xが所属するグループの根を返す
    def root(self, x) -> int:
        if self.parents[x] == x:
            return x
        else:
            self.parents[x] = self.root(self.parents[x])
            return self.parents[x]

    def unite(self, x, y):
        x = self.root(x)
        y = self.root(y)
        if x == y:
            return
        elif x > y:
            u.parents[x] = y
        elif x < y:
            u.parents[y] = x

    def same(self, x, y) -> bool:
        return self.root(x) == self.root(y)


n, m = map(int, input().split())
cnt = n
u = UnionFind(n)
for i in range(m):
    a, b = map(int, input().split())
    if u.same(a-1, b-1):
        continue
    else:
        u.unite(a-1, b-1)
        cnt -= 1
if cnt > 0:
    print(cnt-1)
else:
    print(cnt)
